ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilder.java

```diff
@@ -316,229 +74,166 @@ private static PhysicalSchema getPhysicalSchema(final SourceStep<?> streamSource
+
       if (forceMaterialization) {
         // add this identity mapValues call to prevent the source-changelog
         // optimization in kafka streams - we don't want this optimization to
         // be enabled because we cannot require symmetric serialization between
         // producer and KSQL (see https://issues.apache.org/jira/browse/KAFKA-10179
         // and https://github.com/confluentinc/ksql/issues/5673 for more details)
-        table = source.mapValues(row -> row, materialized);
+        return transformed.mapValues(row -> row, materialized);
+
       } else {
         // if we know this table source is repartitioned later in the topology,
         // we do not need to force a materialization at this source step since the
         // re-partitioned topic will be used for any subsequent state stores, in lieu
         // of the original source topic, thus avoiding the issues above.
         // See https://github.com/confluentinc/ksql/issues/6650
-        table = source.mapValues(row -> row);
+        return transformed.mapValues(row -> row);
       }
-    }
 
-    return table
-        .transformValues(new AddKeyAndTimestampColumns<>(keyGenerator));
   }
 
-  private static StaticTopicSerde.Callback getRegisterCallback(
+  @Override
+  <K> KTable<K, GenericRow> buildWindowedKTable(SourceStep<?> streamSource,
       final RuntimeBuildContext buildContext,
-      final FormatInfo valueFormat
+      final Consumed<K, GenericRow> consumed,
+      final Function<K, Collection<?>> keyGenerator,
+      final Materialized<K, GenericRow, KeyValueStore<Bytes, byte[]>> materialized,
+      final Serde<GenericRow> valueSerde,
+      final String stateStoreName,
+      final PlanInfo planInfo
   ) {
-    final boolean schemaRegistryEnabled = !buildContext
-        .getKsqlConfig()
-        .getString(KsqlConfig.SCHEMA_REGISTRY_URL_PROPERTY)
-        .isEmpty();
-
-    final boolean useSR = FormatFactory
-        .fromName(valueFormat.getFormat())
-        .supportsFeature(SerdeFeature.SCHEMA_INFERENCE);
+    final String changelogTopic = changelogTopic(buildContext, stateStoreName);
+    final Callback onFailure = getRegisterCallback(
+        buildContext, streamSource.getFormats().getValueFormat());
+
+    final KTable<K, GenericRow> source = buildContext
+        .getStreamsBuilder()
+        .table(
+            streamSource.getTopicName(),
+            consumed.withValueSerde(StaticTopicSerde.wrap(changelogTopic, valueSerde, onFailure))
+        );
 
-    if (!schemaRegistryEnabled || !useSR) {
-      return (t1, t2, data) -> { };
-    }
+    final KTable<K, GenericRow> transformed = source
+        .transformValues(new AddKeyAndPseudoColumns<>(
+            keyGenerator, streamSource.getPseudoColumnVersion()));
 
-    return new RegisterSchemaCallback(buildContext.getServiceContext().getSchemaRegistryClient());
+    return transformed.mapValues(row -> row, materialized);
```

Comment:
We don't need a separate `mapValues()` for purposes of materialization. You can materialize on the `transformValues()` directly.